### PR TITLE
Integrate gap junction compensation steps from Oren

### PR DESCRIPTION
**Pull request by WeinaJi**
## Review
* [ ] PR description is complete
* [ ] Coding style (imports, function length, New functions, classes or files) are good
* [ ] Unit/Scientific test added
* [ ] Updated Readme, in-code, developer documentation

### CODE REVIEW

File: tests/scientific/test_simulation.py [hunk 1]
```diff
@@ -121,3 +122,41 @@ def test_v5_gap_junction():
...
+        sim_config_data = json.load(f)
+    sim_config_data["output"]["output_dir"] = "$CURRENT_DIR/output_gj_corrections"
+    sim_config_data["beta_features"] = {
+        "gapjunction_target_population": "default",
+        "determanisitc_stoch": True,
+        "procedure_type": "validation_sim",
+        "gjc": 0.2,
+        "load_g_pas_file": "/gpfs/bbp.cscs.ch/project/proj55/amsalem/gap_junctions/circ19_11_2019_gjs_19_20_20/rm_correction/Circ_mc2_Rt_Remove_ch_all_Det_stoch_True_Dis_holding_False_Correc_type_impedance_tool_Cm0p01_Num_iter_10/num_0/g_pas_passive.hdf5", # noqa
+        "manual_MEComboInfo_file": "/gpfs/bbp.cscs.ch/project/proj55/amsalem/gap_junctions/circ19_11_2019_gjs_19_20_20_new_holding/find_holding_current/vc_ampFile/Circ_mc2_Rt_Remove_ch_none_Det_stoch_True_Dis_holding_False_gjc0p2_Change_mecomb_False_manual_MEComboInfoFile_False_Load_g_pas_True_Correc_iter_loadlast/num_0/holding_per_gid.hdf5" # noqa
+    }
```
Comment (pramodk) [inline tests/scientific/test_simulation.py:144]: @WeinaJi : Just a comment: at a minimum, we should move this to proj12? 

In general, as we want to move the CI out of gitlab / BB5 very soon, it would be good to have tests that are not dependent on BB5/GitLab. Because very soon, we need to work on removing gitlab dependency.

Comment (jorblancoa) [reply tests/scientific/test_simulation.py:144]: this was my comment also, would be nice to avoid depending on bb5 for the tests...

**Comment by jorblancoa** [comment]
General comment:
Do we want to merge those in beta_features? or shall we wait to define them in the sonata-extension and then implement the necessary changes in libsonata first?

**Comment by orena1** [comment]
Hi @WeinaJi I think it looks good, but is there a way to test it?

**Comment by orena1** [comment]
Very cool, a bit off-topic, but there is also the code that generates the GJs and use them. I'll try to get it on github and we can refer to it.

File: tests/scientific/test_simulation.py (hunk 1 above)
Comment (WeinaJi) [reply tests/scientific/test_simulation.py:144]: Yes sure, for this test I will create small, local and dummy calibration files to read as I am not test the final results.

**Comment by WeinaJi** [comment]
> Hi @WeinaJi I think it looks good, but is there a way to test it?

Hello @orena1 , can you access our `bb5` cluster? If yes, you can clone this branch into your directory, load the thalamus module, and overwrite the neurodamus component with your local copy. Something like:
```
git clone git@github.com:BlueBrain/neurodamus.git -b weji/oren
module load unstable neurodamus-thalamus
export PYTHONPATH=$(pwd)/neurodamus:$PYTHONPATH
srun dplace special -mpi -python $NEURODAMUS_PYTHON/init.py --configFile=simulation_config.json
```

The config file `simulation_config.json` should contain the `beta_features` section like
```
{
"beta_features" :{
        "gapjunction_target_population": "thalamus_neurons",
        "determanisitc_stoch": true,
        "procedure_type": "validation_sim",
        "gjc": 0.2,
        "load_g_pas_file": "/gpfs/bbp.cscs.ch/project/proj55/amsalem/gap_junctions/circ19_11_2019_gjs_19_20_20/rm_correction/Circ_mc2_Rt_Remove_ch_all_Det_stoch_True_Dis_holding_False_Correc_type_impedance_tool_Cm0p01_Num_iter_10/num_0/g_pas_passive.hdf5",
        "manual_MEComboInfo_file": "/gpfs/bbp.cscs.ch/project/proj55/amsalem/gap_junctions/circ19_11_2019_gjs_19_20_20_new_holding/find_holding_current/vc_ampFile/Circ_mc2_Rt_Remove_ch_none_Det_stoch_True_Dis_holding_False_gjc0p2_Change_mecomb_False_manual_MEComboInfoFile_False_Load_g_pas_True_Correc_iter_loadlast/num_0/holding_per_gid.hdf5"
    },
}
```

Also cc @PolinaL

**Comment by WeinaJi** [comment]
> Very cool, a bit off-topic, but there is also the code that generates the GJs and use them. I'll try to get it on github and we can refer to it.

That will be great.  Please let me know if you would like to push something in this repo. I can give you the write permission.

In fact, I have another version of your script which includes these commented code:
```
# if load_g_pas_correction_file:
#     logging.info("Loading g_pas_correction_file")
#     exec(open('/gpfs/bbp.cscs.ch/home/amsalem/Dropbox/Blue_Brain/Experiments_Reproduction/Coupling_Coefficient/pylibs/gap_junction_conductance_search_damus.py').read())
```
Is it still needed? I can't find the script `gap_junction_conductance_search_damus.py` any more.

File: neurodamus/gj_user_corrections.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
+# @file gj_user_corrections.py
+# @brief Script for loading user corrections on gap junction connectivity
+# @author Oren Amsalem, Weina Ji
+# @date 2024-09-09
+
+
+import logging
+import numpy as np
+import pickle
+from neurodamus.core import MPI
```
Comment (ferdonline) [inline neurodamus/gj_user_corrections.py:10]:
```suggestion
from .core import MPI
```

File: neurodamus/gj_user_corrections.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+def _perform_remove_channels(node_manager, Mechanisms: list):
+    for cell in node_manager.cells:
+        for sec in cell._cellref.all:
+            for mec in Mechanisms:
+                if mec in dir(sec(.5)): sec.uninsert(mec)
+
+
+def _update_gpas(node_manager, filename, gjc, correction_iteration_load):
+    import h5py
```
Comment (ferdonline) [inline neurodamus/gj_user_corrections.py:114]: nitpick: maybe we would better import h5py at module level. It is not really optional, right?

**Review by ferdonline — DISMISSED** [review]

File: neurodamus/gj_user_corrections.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    processed_cells = 0
+    g_pas_file = h5py.File(filename, 'r')
```
Comment (jorblancoa) [inline neurodamus/gj_user_corrections.py:116]: nitpick:
maybe there should be some error handling here? (same below)
```
try:
    g_pas_file = h5py.File(filename, 'r')
except IOError as e:
    raise ConfigurationError(f"Error opening....: {e}")
```

File: neurodamus/gj_user_corrections.py (hunk 2 above)
Comment (WeinaJi) [reply neurodamus/gj_user_corrections.py]: Done!

File: neurodamus/gj_user_corrections.py (hunk 3 above)
Comment (WeinaJi) [reply neurodamus/gj_user_corrections.py:114]: Not only here, we also import `h5py` at function level in other places in neurodamus because it triggers the numpy warnings. We have a filter to display the warning message once https://github.com/BlueBrain/neurodamus/blob/main/neurodamus/commands.py#L213 but it can only filter when the warning appears at function call.  I remember before I tried to move the filter even earlier in `init.py` but I got some issues with MPI calling before init Neuron.

File: neurodamus/gj_user_corrections.py (hunk 4 above)
Comment (WeinaJi) [reply neurodamus/gj_user_corrections.py]: Good point. I've added it.

**Review by jorblancoa — APPROVED** [review]

File: neurodamus/gap_junction.py [hunk 5]
```diff
@@ -66,6 +69,8 @@ def __init__(self, gj_conf, target_manager, cell_manager, src_cell_manager=None,
         super().__init__(gj_conf, target_manager, cell_manager, src_cell_manager, **kw)
         self._src_target_filter = target_manager.get_target(cell_manager.circuit_target,
                                                             src_cell_manager.population_name)
+        self.holding_ic_per_gid = None
+        self.SEClamp_current_per_gid = None
```
Comment (ferdonline) [inline neurodamus/gap_junction.py:73]: totally nitpick, sorry, but would it be too bad to adhere to normal lowercase?
```suggestion
        self.seclamp_current_per_gid = None
```

File: neurodamus/gap_junction.py [hunk 6]
```diff
@@ -98,6 +103,10 @@ def configure_connections(self, conn_conf):
 
     def finalize(self, *_, **_kw):
         super().finalize(conn_type="Gap-Junctions")
+        if self.cell_manager.population_name == \
+                SimConfig.beta_features.get("gapjunction_target_population"):
```
Comment (ferdonline) [inline neurodamus/gap_junction.py:107]: Just another style nitpick 🙈 
```suggestion
        gj_target_pop = SimConfig.beta_features.get("gapjunction_target_population")
        if self.cell_manager.population_name == gj_target_pop:
```

File: neurodamus/gj_user_corrections.py [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+    Mechanisims = []
+    if remove_channels == 'all':
+        Mechanisims = non_stochastic_mechs + stochastic_mechs
+    if remove_channels == 'only_stoch':
+        Mechanisims = stochastic_mechs
+    if remove_channels == 'only_non_stoch':
+        Mechanisims = non_stochastic_mechs
+    if remove_channels:
+        logging.info("Removing channels type = " + remove_channels)
+        _perform_remove_channels(node_manager, Mechanisims)
```
Comment (ferdonline) [inline neurodamus/gj_user_corrections.py:52]: Unnecessary checks, var naming typos, missing warning
```suggestion
    if remove_channels:
        if remove_channels == 'all':
            rm_mechanisms = non_stochastic_mechs + stochastic_mechs
        elif remove_channels == 'only_stoch':
            rm_mechanisms = stochastic_mechs
        elif remove_channels == 'only_non_stoch':
            rm_mechanisms = non_stochastic_mechs
        else:
            logging.warning("Unknown GJ remove_channels setting: %s", remove_channels)
            rm_mechanisms = []
        logging.info("Removing channels type = " + remove_channels)
        _perform_remove_channels(node_manager, rm_mechanisms)
```

**Review by ferdonline — COMMENTED** [review]
Some nitpicks, but it might be worth to a wider round of checks for naming and other small inconsistencies. vars starting in upcase remember class variables or constants

**Comment by WeinaJi** [comment]
> Very cool, a bit off-topic, but there is also the code that generates the GJs and use them. I'll try to get it on github and we can refer to it.

Hello @orena1, have you managed to test this branch on bb5? Or would you like to add your code in it?
As BBP closure is approaching, I would like to merge this feature into the main branch although it is still in "beta_futures".  In this way, @PolinaL could check it out in another computing cluster after bb5 to continue her simulation.

File: neurodamus/gj_user_corrections.py (hunk 7 above)
Comment (WeinaJi) [reply neurodamus/gj_user_corrections.py]: Done!

File: neurodamus/gap_junction.py (hunk 6 above)
Comment (WeinaJi) [reply neurodamus/gap_junction.py]: Thanks, done!

File: neurodamus/gap_junction.py (hunk 5 above)
Comment (WeinaJi) [reply neurodamus/gap_junction.py]: done

**Review by ferdonline — DISMISSED** [review]

**Review by jorblancoa — DISMISSED** [review]

**Review by jorblancoa — APPROVED** [review]

File: neurodamus/gap_junction.py [hunk 8]
```diff
@@ -98,6 +103,10 @@ def configure_connections(self, conn_conf):
 
     def finalize(self, *_, **_kw):
         super().finalize(conn_type="Gap-Junctions")
+        if (gj_target_pop := SimConfig.beta_features.get("gapjunction_target_population")) \
```
[the remark this answers is not in the document]
Comment (WeinaJi) [reply neurodamus/gap_junction.py:106]: `gj_target_pop` is the population name. We need to test it not `None`, i.e. has been set in "beta_feature".

**Comment by orena1** [comment]
Hi, sorry for the late response, this file is probably relevant for the process:
https://github.com/orena1/Experiments_Reproduction/blob/master/Coupling_Coefficient/send_sim_to_share_v1.py
